packages/history/index.ts

```diff
@@ -685,15 +685,16 @@ export function createHashHistory(
   }
 
   function createHref(to: To) {
-    return getBaseHref() + '#' + (typeof to === 'string' ? to : createPath(to));
+    let path = resolvePath(to, hashRoot)
+    return getBaseHref() + '#' + createPath(path);
   }
 
   function getNextLocation(to: To, state: any = null): Location {
     return readOnly<Location>({
       pathname: location.pathname,
       hash: '',
       search: '',
-      ...(typeof to === 'string' ? parsePath(to) : to),
+      ...resolvePath(to, hashRoot),
```

Comment:
This is the line that’s required to support relative paths in the hash state for `hashRoot=“”`.